Supervisor 3.2.1 (new formula)

MerelyAPseudonym opened the pull request:
### All Submissions:

- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew/pulls) for the same update/change?

### New Formulae Submissions:

- [X] Does your submission pass
`brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Have you built your formula locally prior to submission with `brew install <formula>`?

Kudos to @deployable for getting the ball rolling with [#38552](https://github.com/Homebrew/homebrew/pull/38552)!

DomT4 commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,108 @@
+  url "https://github.com/Supervisor/supervisor/archive/3.2.1.tar.gz"
+  sha256 "d9c0b17ec42ac6477e7ef0ad2e1f6f1597f855a4d9606c024ac40eb034d6e9ed"
+
+  resource "meld3" do
```
You might need a:
```ruby
depends_on :python if MacOS.version <= :snow_leopard
```
in this formula line as well.

MerelyAPseudonym replied:
Is it still necessary even if Supervisor will work on Python 2.5+? (and actually on Python 2.4 too if the [ElementTree](http://effbot.org/zone/element-index.htm) library is installed)

DomT4 replied:
No. If everything about `supervisor` and `meld3` can run on ancient Pythons it's fine to leave the line out.

tdsmith replied:
It's necessary because you use literal `2.7`'s below.

DomT4 replied:
TIL. Thanks Tim :)

MerelyAPseudonym replied:
D'oh! Nice catch @tdsmith

DomT4 commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,108 @@
+    # it to use Homebrew's conventions.
+    inreplace buildpath/"supervisor/skel/sample.conf" do |s|
+      # A `[unix_http_server]` modification
+      s.gsub! %r{(?<=^file=)/tmp/supervisor.sock}, var/"run/supervisor.sock"
```
You'll want to create these `var` directories in `post_install`.

MerelyAPseudonym replied:
Nice catch! Will do.

MerelyAPseudonym replied:
@DomT4 If the directories already exist I'm seeing a warning. Should I add something like `unless the_path.exist?`to the `mkdir` lines?

MerelyAPseudonym replied:
The warnings look like:

```
/usr/local/Library/Homebrew/postinstall.rb (Formulary::FromPathLoader): loading /usr/local/Library/Formula/supervisor.rb
Warning: The post-install step did not complete successfully
You can try again using `brew postinstall supervisor`
==> File exists - /usr/local/var/run
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:245:in `mkdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:245:in `fu_mkdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:174:in `block in mkdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:173:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/fileutils.rb:173:in `mkdir'
/usr/local/Library/Homebrew/extend/fileutils.rb:52:in `mkdir'
/usr/local/Library/Formula/supervisor.rb:86:in `post_install'
/usr/local/Library/Homebrew/formula.rb:822:in `run_post_install'
/usr/local/Library/Homebrew/postinstall.rb:16:in `<main>'
```

DomT4 commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,108 @@
+
+    bin.install Dir[libexec/"bin/*"]
+
+    ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python2.7/site-packages"
```
It's a style nit but these are usually above the respective installs, i.e. the vendor one above the resource install and the lib one above the main compile.

MerelyAPseudonym replied:
Yeah, these `PYTHONPATH` lines are so far down because I was exploring what was actually *required* for this formula for build correctly. So, while I think it's notable that the resource install and main compile succeeded without the `PYTHONPATH` environment tweaks, I'm happy to defer to the Homebrew conventions.

DomT4 commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+  end
+
+  plist_options :manual => "supervisord -c #{HOMEBREW_PREFIX}/etc/supervisord.ini"
```
Mind moving this and the `plist` above the test? Ideally `test do` should be the final block.

MerelyAPseudonym replied:
I don't mind at all. Is that something that `brew audit` should check for?

DomT4 replied:
It sort-of already is, but we haven't taught the [audit](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/audit.rb#L170-L191) that `plist`s are a thing yet so it ignores the technical violation heh.

DomT4 commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,108 @@
+      <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+      <plist version="1.0">
+        <dict>
+
```
Scrap the empty lines please. They're not that common in plists and I'm not convinced it makes it significantly easier to read to justify the generous spacing.

deployable commented:
Thanks for the update @MerelyAPseudonym, perfect timing.

MerelyAPseudonym commented:
Sorry for the delay everyone!

For ease of review, I've addressed the PR comments in distinct commits on my original branch. If/When everything's approved, I'll rebase onto `master`, squashing back to the original two commits.

MerelyAPseudonym commented:
Heads up: there might be one more issue with the test and the sandbox, but since I'm having trouble reproducing it I'll just wait to see how Travis fares.

mikemcquaid commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,117 @@
+  sha256 "d9c0b17ec42ac6477e7ef0ad2e1f6f1597f855a4d9606c024ac40eb034d6e9ed"
+
+  # Although Supervisor itself doesn't require Python 2.7 (its actual lower
+  # bound is 2.4) this formula only targets 2.7 for simplicity.
```
You can skip these comments: it's implied by the use of `2.7` frequently below.

mikemcquaid commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,117 @@
+  def post_install
+    # If the directories that supervisord expects to place its runtime files
+    # into don't exist, it dies instead of creating them itself.
+    mkdir var/"run" unless (var/"run").exist?
```
Just use `(var/"run").mkpath` and skip the `.exist?` (and below)

MerelyAPseudonym replied:
Ahh, `.mkpath`, **that's** what I needed! =)

Thanks for the tip.

mikemcquaid commented:
You can scrap all (or pretty much all) the comments in this formula as the code is self-explanatory.

MerelyAPseudonym commented:
@mikemcquaid yessir

MerelyAPseudonym commented:
Alright, I think everything's good to go. I can rebase-&-squash upon request.

MerelyAPseudonym commented:
(Or address further concerns or suggestions, of course)

huguesalary commented:
When install supervisord with this formula, at the end of the `supervisord.ini` file you can read:
```
[include]
files = /usr/local/etc/supervisor.d/.ini
```

This line should instead read:
```
[include]
files = /usr/local/etc/supervisor.d/*.ini
```

Note the `*.ini`. 

Without this change, files in `/usr/local/etc/supervisor.d/` are not loaded. I, however, have a feeling that this is something to fix upstream.

MerelyAPseudonym commented:
@huguesalary You're absolutely right that the line should read `files = /usr/local/etc/supervisor.d/*.ini`. However, it's not an upstream issue—it was just [a goof] on my part. I appreciate your calling it out, despite it making me look bad ;P

[a goof]: https://github.com/MerelyAPseudonym/homebrew/commit/84d1a5c194daed113b9213c476296ced2b605b58#diff-accbf3783802d4399fab1d36d295f198R33

mikemcquaid commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,113 @@
+
+    bin.env_script_all_files(libexec/"bin", :PYTHONPATH => ENV["PYTHONPATH"])
+
+    # Q.v. <https://github.com/Supervisor/supervisor/issues/608>.
```
Not sure what `Q.v` means here; can you elaborate a bit in the comment? Also, I'd avoid the `<..>` wrapping. Thanks!

MerelyAPseudonym replied:
Hm, yeah, this should be made less cryptic.

If you were interested: “q.v.” stands for <i>quod vidē</i>, and it’s used to point to a reference or an example.

mikemcquaid commented on Library/Formula/supervisor.rb:
```diff
@@ -0,0 +1,113 @@
+
+  def install
+    inreplace buildpath/"supervisor/skel/sample.conf" do |s|
+      # A `[unix_http_server]` modification
```
Still think you can kill all these comments. Additionally, looks like you could just do a single global replace on `/tmp/supervisord\.`

MerelyAPseudonym replied:
I'm not used to being told to *remove* comments lol

Anyway, will do!

mikemcquaid commented:
@MerelyAPseudonym Almost there; a few final comments but looking good :+1:

MerelyAPseudonym commented:
Thanks for the encouragement @mikemcquaid! I just pushed a couple more commits based on your recent comments.

mikemcquaid commented:
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!
